src/scvi/external/methylvi/_model.py

```diff
@@ -0,0 +1,704 @@
+    def setup_mudata(
+        cls,
+        mdata: MuData,
+        mc_layer: str,
+        cov_layer: str,
+        methylation_modalities: Iterable[str],
+        batch_key: str | None = None,
+        covariate_modalities=None,
+        **kwargs,
+    ):
+        """%(summary_mdata)s.
+
+        Parameters
+        ----------
+        %(param_mdata)s
+        mc_layer
+            Layer containing methylated cytosine counts for each set of methylation features.
+        cov_layer
+            Layer containing total coverage counts for each set of methylation features.
+        methylation_modalities
+            List of modalities in `mdata` object representing different methylation contexts.
+            Each modality must be equipped with a layer containing the number of methylated counts
+            (specified by `mc_layer`) and total number of counts (specified by `cov_layer`) for
+            each genomic region feature.
+        %(param_batch_key)s
+        %(param_covariate_modalities)s
+
+        Examples
+        --------
+        MethylVI.setup_mudata(
+            mdata,
+            mc_layer="mc",
+            cov_layer="cov",
+            batch_key="Platform",
+            methylation_modalities=['mCG', 'mCH'],
+            covariate_modalities={
+                "batch_key": "mCG"
+            },
+        )
+
+        """
+        if covariate_modalities is None:
+            covariate_modalities = {}
+        setup_method_args = MethylVIModel._get_setup_method_args(**locals())
+
+        if methylation_modalities is None:
+            raise ValueError("Methylation modalities cannot be None.")
+
+        covariate_modalities_ = cls._create_modalities_attr_dict(
+            covariate_modalities, setup_method_args
+        )
+
+        batch_field = fields.MuDataCategoricalObsField(
+            REGISTRY_KEYS.BATCH_KEY,
+            batch_key,
+            mod_key=covariate_modalities_.batch_key,
+        )
+
+        mc_fields = []
+        cov_fields = []
+
+        for mod in methylation_modalities:
+            mc_fields.append(
+                fields.MuDataLayerField(
+                    f"{mod}_{METHYLVI_REGISTRY_KEYS.MC_KEY}",
+                    mc_layer,
+                    mod_key=mod,
+                    is_count_data=True,
+                    mod_required=True,
+                )
+            )
+
+            cov_fields.append(
+                fields.MuDataLayerField(
+                    f"{mod}_{METHYLVI_REGISTRY_KEYS.COV_KEY}",
+                    cov_layer,
+                    mod_key=mod,
+                    is_count_data=True,
+                    mod_required=True,
+                )
+            )
+
+        mudata_fields = mc_fields + cov_fields + [batch_field]
+        adata_manager = AnnDataManager(fields=mudata_fields, setup_method_args=setup_method_args)
+        adata_manager.register_fields(mdata, **kwargs)
+        adata_manager.modalities = methylation_modalities
```

Comment:
This (and a similar line in `setup_anndata`) was a hack so that I could access modality names later when instantiating the class. Is there a smarter way to do this?

Reply:
I guess the correct way would be to iterate over the layer fields and split the "mod" from above. I don't like treating adata_manager as a bucket. You can have a small helper function that takes care of this before setting up the model and stores it under model.modalities.

Reply:
I agree that this initial solution wasn't great. To resolve this, I accessedthe anndata manager within the model's `__init__` function and then retrieved the setup method args to get the modality/context names (see line 84 in the newest version of `_model.py`). Let me know what you think of this solution.